File: scripts/test_solutions/131_palindrome_partitioning.js
var partition = function(s) {
    const result = [];
    function isPalin(l, r) { while (l < r) if (s[l++] !== s[r--]) return false; return true; }
    function backtrack(start, path) {
        if (start === s.length) { result.push([...path]); return; }
        for (let end = start; end < s.length; end++) {
            if (isPalin(start, end)) {
                path.push(s.substring(start, end + 1));
                backtrack(end + 1, path);
                path.pop();
            }
        }
    }
    backtrack(0, []);
    return result;
};

File: scripts/test_solutions/131_palindrome_partitioning.py
def partition(s):
    result = []

    def is_palin(l, r):
        while l < r:
            if s[l] != s[r]:
                return False
            l += 1
            r -= 1
        return True

    def backtrack(start, path):
        if start == len(s):
            result.append(path[:])
            return
        for end in range(start, len(s)):
            if is_palin(start, end):
                path.append(s[start:end + 1])
                backtrack(end + 1, path)
                path.pop()

    backtrack(0, [])
    return result
